File: backend/src/main/java/com/senior/project/backend/domain/Task.java
package com.senior.project.backend.domain;

import jakarta.persistence.*;
import lombok.*;

@Getter
@Setter
@Entity
@AllArgsConstructor
@NoArgsConstructor
@Generated
public class Task {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    private String name;
    public String description;
    private Boolean isRequired;
    @Enumerated(EnumType.STRING)
    private YearLevel yearLevel;
    @Enumerated(EnumType.STRING)
    private TaskType taskType;
    private String artifactName;
    private String submissionInstructions;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn()
    private Milestone milestone;

    @OneToOne(fetch = FetchType.EAGER)
    @JoinColumn
    private Event event;
}

File: frontend/crd/src/domain/Task.ts
import { YearLevel } from "./Milestone";

export enum TaskType {
  ARTIFACT = 'ARTIFACT',
  COMMENT = 'COMMENT',
  EVENT = 'EVENT'
}

export interface TaskJSON {
  name: string;
  description: string;
  id: number;
  isRequired: boolean;
  yearLevel: YearLevel;
  milestoneID?: number;
  taskType: string;
  artifactName?: string;
  eventID?: number;
  submissionInstructions?: string;
}

export class Task {
  constructor(json: TaskJSON) {
    this.name = json.name;
    this.description = json.description;
    this.taskID = json.id;
    this.isRequired = json.isRequired;
    this.yearLevel = json.yearLevel;
    this.milestoneID = json?.milestoneID;
    this.taskType = json.taskType;
    this.artifactName = json?.artifactName;
    this.eventID = json?.eventID;
    this.submissionInstructions = json?.submissionInstructions;
  }

  name: string;
  description: string;
  taskID: number;
  isRequired: boolean;
  yearLevel: YearLevel;
  milestoneID?: number;
  taskType: string;
  artifactName?: string;
  eventID?: number;
  submissionInstructions?: string;

  needsArtifact(): boolean {
    return this.taskType === TaskType.ARTIFACT;
  }
}
